scrape/scrape.go

```diff
@@ -413,31 +421,50 @@ func (sp *scrapePool) Sync(tgs []*targetgroup.Group) {
 	start := time.Now()
 
 	sp.targetMtx.Lock()
-	var all []*Target
-	sp.droppedTargets = []*Target{}
-	for _, tg := range tgs {
-		targets, err := targetsFromGroup(tg, sp.config)
-		if err != nil {
-			level.Error(sp.logger).Log("msg", "creating targets failed", "err", err)
-			continue
-		}
-		for _, t := range targets {
-			if t.Labels().Len() > 0 {
-				all = append(all, t)
-			} else if t.DiscoveredLabels().Len() > 0 {
-				sp.droppedTargets = append(sp.droppedTargets, t)
-			}
-		}
-	}
+	targets := sp.buildTargetsToSync(tgs)
 	sp.targetMtx.Unlock()
-	sp.sync(all)
+
+	sp.sync(targets)
 
 	targetSyncIntervalLength.WithLabelValues(sp.config.JobName).Observe(
 		time.Since(start).Seconds(),
 	)
 	targetScrapePoolSyncsCounter.WithLabelValues(sp.config.JobName).Inc()
 }
 
+func (sp *scrapePool) buildTargetsToSync(tgs []*targetgroup.Group) []*Target {
+	all := make([]*Target, 0)
+	dropped := make(map[uint64]*Target)
+	for _, tg := range tgs {
+		for _, tlset := range tg.Targets {
+			origLabels := stackTargetLabels(tlset, tg.Labels, sp.config)
+
```

Comment:
We need a comment somewhere around here saying that we want to avoid the cost of doing relabelling for targets we know will be dropped.